pkg/telemetry/events.go

```diff
@@ -145,6 +145,33 @@ func (t *telemetryService) ParticipantResumed(
 	reason livekit.ReconnectReason,
 ) {
 	t.enqueue(func() {
+		// create a worker if needed.
+		//
+		// Signalling channel stats collector and media channel stats collector could both call
+		// ParticipantJoined and ParticipantLeft.
+		//
+		// On a resume, the signalling channel collector would call `ParticipantLeft` which would close
+		// the corresponding participant's stats worker.
+		//
+		// So, on a successful resume, create the worker if needed.
+		worker, found := t.getOrCreateWorker(
+			ctx,
+			livekit.RoomID(room.Sid),
+			livekit.RoomName(room.Name),
+			livekit.ParticipantID(participant.Sid),
+			livekit.ParticipantIdentity(participant.Identity),
+		)
+		if !found {
+			// a worker was created, check if there is a closed worker and transfer connected state to new worker
+			if closedWorker := t.getClosedWorker(livekit.ParticipantID(participant.Sid)); closedWorker != nil {
```

Comment:
it is not possible to enter this branch. `getorCreateWorker` replaces the old worker in the worker map so when `getClosedWorker` is called it will return nil because the new worker isn't closed. if we only want to do this connected state transfer from resume maybe we could add `copyConnectedState` or something like that as an arg to `getOrCreateWorker` to instruct it to do this if it replaces a closed worker.

Reply:
Oh good catch.

I will actually change the `getClosedWorker` to walk the list and find a closed one? `copyConnectedState` is better though as it will be atomic, but was thinking about adding an argument everywhere. Actually, it is better. I will add that argument.

Reply:
Changed in this [commit](https://github.com/livekit/livekit/pull/2982/commits/214e5ca055b077b37fd278deb0ef9c4136c67f50). It was not bad. There were only three call sites. Thought there were more.

Reply:
for context - only access to the head of `workerList` is protected by `workersMu` - the rest of the links are protected by `flushMu` so that flushing and other ops don't block each other. `workersMu` is locked while holding `flushMu` so this might get tricky if we needed to traverse the list in other places

Reply:
Yeah, not a good idea to traverse other than in `Flush`. Use the extra argument like you suggested.